// @angular/core/esm2022/src/initial_render_pending_tasks.mjs
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { Injectable } from './di';
import { inject } from './di/injector_compatibility';
import { NgZone } from './zone/ng_zone';
import * as i0 from "./r3_symbols";
/**
 * *Internal* service that keeps track of pending tasks happening in the system
 * during the initial rendering. No tasks are tracked after an initial
 * rendering.
 *
 * This information is needed to make sure that the serialization on the server
 * is delayed until all tasks in the queue (such as an initial navigation or a
 * pending HTTP request) are completed.
 */
class InitialRenderPendingTasks {
    get whenAllTasksComplete() {
        if (this.collection.size === 0) {
            this.complete();
        }
        return this.promise;
    }
    constructor() {
        this.taskId = 0;
        this.collection = new Set();
        this.ngZone = inject(NgZone);
        this.completed = false;
        // Run outside of the Angular zone to avoid triggering
        // extra change detection cycles.
        this.ngZone.runOutsideAngular(() => {
            this.promise = new Promise((resolve) => {
                this.resolve = resolve;
            });
        });
    }
    add() {
        if (this.completed) {
            // Indicates that the task was added after
            // the task queue completion, so it's a noop.
            return -1;
        }
        const taskId = this.taskId++;
        this.collection.add(taskId);
        return taskId;
    }
    remove(taskId) {
        if (this.completed)
            return;
        this.collection.delete(taskId);
        if (this.collection.size === 0) {
            this.complete();
        }
    }
    ngOnDestroy() {
        this.complete();
        this.collection.clear();
    }
    complete() {
        this.completed = true;
        this.resolve();
    }
    static { this.ɵfac = function InitialRenderPendingTasks_Factory(t) { return new (t || InitialRenderPendingTasks)(); }; }
    static { this.ɵprov = /*@__PURE__*/ i0.ɵɵdefineInjectable({ token: InitialRenderPendingTasks, factory: InitialRenderPendingTasks.ɵfac, providedIn: 'root' }); }
}
export { InitialRenderPendingTasks };
(function () { (typeof ngDevMode === "undefined" || ngDevMode) && i0.setClassMetadata(InitialRenderPendingTasks, [{
        type: Injectable,
        args: [{ providedIn: 'root' }]
    }], function () { return []; }, null); })();
